kernel/work_q.c

```diff
@@ -147,4 +147,82 @@ bool k_delayed_work_pending(struct k_delayed_work *work)
 	       k_work_pending(&work->work);
 }
 
+static void work_delayable_timeout(struct _timeout *t)
+{
+	struct k_work_delayable *w = CONTAINER_OF(t, struct k_work_delayable,
+						   timeout);
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	struct k_work_q *wq = w->work_q;
+
+	w->work_q = NULL;
+	k_spin_unlock(&lock, key);
+
+	/* submit work to workqueue */
+	k_work_submit_to_queue(wq, &w->work);
+}
+
+void k_work_delayable_init(struct k_work_delayable *work, k_work_handler_t handler)
+{
+	k_work_init(&work->work, handler);
+	z_init_timeout(&work->timeout);
+	work->work_q = NULL;
+}
+
+int k_work_schedule_for_queue(struct k_work_q *work_q,
+			      struct k_work_delayable *work,
+			      k_timeout_t delay)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	int err = 0;
+
+	/* Work cannot be active in multiple queues */
+	if (work->work_q != NULL && work->work_q != work_q) {
+		err = -EBUSY;
+		goto unlock;
+	}
+
+	/* Cancel any incomplete timeout. */
+	if (work->work_q == work_q) {
+		z_abort_timeout(&work->timeout);
+		work->work_q = NULL;
+	}
+
+	/* If the timeout is no-wait, just submit it.  If it's already
+	 * pending its position will be unchanged.
+	 */
+	if (K_TIMEOUT_EQ(delay, K_NO_WAIT)) {
+		k_spin_unlock(&lock, key);
+		k_work_submit_to_queue(work_q, &work->work);
```

Comment:
submitting outside a lock can lead to a race with cancel
if cancel is called after mutex is release but before actual submission it will return `-EINVAL` while it should return `-EPENDING`

Reply:
At least I think so as from the functional point of view the work is "pending" and there is no way of not invoking this submit.
On the other hand the function has not returned yet and such race is indistinguishable from a case when we have cancel just before the submit so cancel should not be called this way to obtain the right value.

Reply:
> so cancel should not be called this way to obtain the right value.

Yes: that's the fundamental flaw with the way the current API is being used.  Basically anything that's **depending** on the value of `k_work_pending()` to affect control flow is fragile.

I'll update the documentation to be a bit more clear.